In [61]:
#Import Libraries
import pandas as pd
import numpy as np
import os


In [62]:
path=r'C:\Users\User\OneDrive\Belgeler\07-02-24 Instacart Basket Analysis'

In [63]:
path

'C:\\Users\\User\\OneDrive\\Belgeler\\07-02-24 Instacart Basket Analysis'

In [64]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products[1].csv'), index_col = False)

In [65]:
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [66]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   time_of_placing_order   int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [67]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   time_of_placing_order  days_since_prior_order  
0                      8                     NaN  
1                      7                    15.0  
2                     12                    21.0  
3                      7                    29.0  
4                     15                    28.0

In [68]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
time_of_placing_order          0
days_since_prior_order    206209
dtype: int64

In [69]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  time_of_placing_order  days_since_prior_order  
count        3.421083e+06           3.421083e+06            3.214874e+06  
mean         2.776219e+00           1.345202e+01            1.111484e+01  
std          2.046829e+00           4.226088e+00            9.206737e+00  
min          0.000000e+00           0.000000e+00            0.000000e+00  
25%          1.000000e+00           1.000000e+01            4.000000e+00  
50%          3.000000e+00           1.300000e+01            7.000000e+00  
75%          5.000000e+00           1.600000e+01            1.500000e+01  
max          6.000000e+00           2.300000e+01            3.000000e+01

In [70]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   time_of_placing_order  days_since_prior_order  
0                      8                     NaN  
1                      7                    15.0  
2                     12                    21.0  
3                      7                    29.0  
4                     15                    28.0

In [71]:
#Create a dataframe
df_test=pd.DataFrame

In [72]:

df_test['mix'] = ['a', 'b', 1, True]

TypeError: 'type' object does not support item assignment

In [73]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [74]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [75]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [76]:
df_prods.shape

(49693, 5)

In [77]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [78]:
df_prods_clean.shape

(49677, 5)

In [79]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [80]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [81]:
df_prods_clean.shape

(49677, 5)

In [82]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [83]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5: Data Consistency Checks


#2.Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [84]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  time_of_placing_order  days_since_prior_order  
count        3.421083e+06           3.421083e+06            3.214874e+06  
mean         2.776219e+00           1.345202e+01            1.111484e+01  
std          2.046829e+00           4.226088e+00            9.206737e+00  
min          0.000000e+00           0.000000e+00            0.000000e+00  
25%          1.000000e+00           1.000000e+01            4.000000e+00  
50%          3.000000e+00           1.300000e+01            7.000000e+00  
75%          5.000000e+00           1.600000e+01            1.500000e+01  
max          6.000000e+00           2.300000e+01            3.000000e+01

#Orders_day_of_ week column shows results in a min of 0 and a max of 6. 50th percentile is 3.
#There is not  a negative min value for days_since_prior_order.
#Time_of_placing_order column has a max value shorter than 24 hours.

#3.Check for mixed-type data in your df_ords dataframe.

In [85]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#There is no mixed-type data in df_ords dataframe.

In [86]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
time_of_placing_order       int64
days_since_prior_order    float64
dtype: object

#5 Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [87]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
time_of_placing_order          0
days_since_prior_order    206209
dtype: int64

#It seems that the column "days_since_prior_order" has 206,209 missing values (NaNs). This column likely represents the number of days since the prior order for each order in the dataset. The missing values indicate either that there were no prior orders for those users or that the data is unavailable for some reason. 

In [88]:
df_ords.shape

(3421083, 7)

In [89]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   time_of_placing_order   int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


#6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [90]:
df_ords.median()

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
time_of_placing_order          13.0
days_since_prior_order          7.0
dtype: float64

In [91]:
df_ords.mean()

Unnamed: 0                1.710541e+06
order_id                  1.710542e+06
user_id                   1.029782e+05
order_number              1.715486e+01
orders_day_of_week        2.776219e+00
time_of_placing_order     1.345202e+01
days_since_prior_order    1.111484e+01
dtype: float64

In [92]:
median_value = df_ords['days_since_prior_order'].median()

In [93]:
# I would choose median because placing order once a week makes more sense than placing every other day.
df_ords['days_since_prior_order'].fillna(median_value, inplace=True)

In [94]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   time_of_placing_order  days_since_prior_order  
0                      8                     7.0  
1                      7                    15.0  
2                     12                    21.0  
3                      7                    29.0  
4                     15                    28.0

#7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [95]:
df_dups_ords=df_ords[df_ords.duplicated()]

In [96]:
df_ords.shape

(3421083, 7)

In [97]:
df_dups_ords.shape

(0, 7)

Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.
Answer:There is no duplicates.

In [98]:
df_ords_no_dups=df_ords.drop_duplicates()

In [99]:
df_ords_no_dups

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         time_of_placing_order  days_since_prior_order  
0                            8                     7.0  
1                            7                    15.0  
2                           12                    21.0  
3                            7                    29.0  
4                           15                    28.0  
...                        ...                     ...  
3421078                     18                    29.0  
3421079                     10                    30.0  
3421080                     12                    18.0  
3421081                     12                     7.0  
3421082                     14                    30.0  

[3421083 rows x 7 columns]

In [100]:
df_ords_no_dups=df_ords

In [101]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   time_of_placing_order  days_since_prior_order  
0                      8                     7.0  
1                      7                    15.0  
2                     12                    21.0  
3                      7                    29.0  
4                     15                    28.0

In [102]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'),index = False)

In [103]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'),index = False)